projects/knora-ui/src/lib/viewer/values/link-value/link-value.component.ts

```diff
@@ -0,0 +1,151 @@
+    } else {
+      return null;
+    }
+  }
+  // override the resetFormControl() from the base component to deal with initial link value label
+  resetFormControl(): void {
+    super.resetFormControl();
+
+    if (this.valueFormControl !== undefined) {
+      if (this.mode === 'read') {
+
+        const initialValue = this.getInitValue();
+        const initialComment = this.getInitComment();
+        this.valueFormControl.setValue(initialValue.label);
+        this.commentFormControl.setValue(initialComment);
+
+        this.valueFormControl.clearValidators();
+      } else {
+        this.valueFormControl.setValue('');
+        this.valueChangesSubscription = this.valueFormControl.valueChanges.subscribe(data => {
+          this.options = this.searchByLabel(data);
+        });
+        this.valueFormControl.clearValidators();
+      }
+    }
+  }
+
+  ngOnInit() {
+    // initialize form control elements
+    this.valueFormControl = new FormControl(null);
+
+    this.commentFormControl = new FormControl(null);
+
+    // subscribe to any change on the comment and recheck validity
+    this.valueChangesSubscription = this.commentFormControl.valueChanges.subscribe(
+      data => {
+        this.valueFormControl.updateValueAndValidity();
+      }
+    );
+
+    this.form = this.fb.group({
+      linkValue: this.valueFormControl,
+      comment: this.commentFormControl
+    });
+    this.options = [this.displayValue.linkedResource];
+    this.resetFormControl();
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.resetFormControl();
+  }
+
+  // unsubscribe when the object is destroyed to prevent memory leaks
+  ngOnDestroy(): void {
+    this.unsubscribeFromValueChanges();
+  }
+  getNewValue(): CreateLinkValue | false {
+    if (this.mode !== 'create' || !this.form.valid) {
+      return false;
+    }
+    const newLinkValue = new CreateLinkValue();
+    // check that resource with given label exists
+    newLinkValue.linkedResourceIri = this.linkedResourceIRI;
+
+    if (this.commentFormControl.value !== null && this.commentFormControl.value !== '') {
+      newLinkValue.valueHasComment = this.commentFormControl.value;
+    }
+
+    return newLinkValue;
+  }
+
+  getUpdatedValue(): UpdateLinkValue | false {
+    if (this.mode !== 'update' || !this.form.valid) {
+      return false;
+    }
+    const updatedLinkValue = new UpdateLinkValue();
+    updatedLinkValue.id = this.displayValue.id;
+    updatedLinkValue.linkedResourceIri = this.linkedResourceIRI;
+    // add the submitted comment to updatedLinkValue only if user has added a comment
+    if (this.commentFormControl.value !== null && this.commentFormControl.value !== '') {
+      updatedLinkValue.valueHasComment = this.commentFormControl.value;
+    }
+
+    return updatedLinkValue;
+  }
+  onOptionChange(option: ReadResource) {
+      this.setLinkValue(option);
+  }
+
+  setLinkValue(option: ReadResource) {
+    this.valueFormControl.setValue(option.label);
```

Comment:
This is not how it should be done because the flow of date is not solely controlled by `FormControl`. Two additional members `linkedResourceIRI` and `linkedResource` are required to manage the state. Maybe making the integer the reference component was not ideal since a number is primitive. It gets more complex when dealing with objects like in case of the interval value.

I see two options:

1. make the `FormControl`'s value a `ReadResource` and use autocomplete's `displayWith` input to show the label to the user (https://material.angular.io/components/autocomplete/overview#setting-separate-control-and-display-values). Then get rid of the additional class members, you can access the `ReadResource` via the `FormControl`. See https://github.com/dasch-swiss/knora-ui/blob/master/projects/knora/search/src/lib/extended-search/select-property/specify-property-value/link-value/link-value.component.html.

2. if 1. does not do the job, you can always make a custom subcomponent as we did for the interval or color value. However, I think 1. is preferable since it seems simpler to me.

If you need any help, please let me know.

Reply:
thanks for this, option 1 works.